components/patient-data/impl/src/main/java/org/phenotips/data/internal/PatientOnCancelDeleter.java

```diff
@@ -0,0 +1,124 @@
+import org.slf4j.Logger;
+
+import com.xpn.xwiki.XWiki;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.criteria.impl.RevisionCriteria;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.doc.rcs.XWikiRCSNodeInfo;
+
+/**
+ * Deletes the patient record when a {@link PatientEditingCancelingEvent} is received from a new patient record.
+ *
+ * @since 1.2RC1
+ * @version $Id$
+ */
+@Component
+@Named("patient-oncancel-deleter")
+@Singleton
+public class PatientOnCancelDeleter extends AbstractEventListener
+{
+    private static final String AUTOSAVE_COMMENT = "(Autosaved)";
+
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    @Inject
+    private Logger logger;
+
+    /** Fills in missing reference fields with those from the current context document to create a full reference. */
+    @Inject
+    @Named("current")
+    private EntityReferenceResolver<EntityReference> currentResolver;
+
+
+    /** Default constructor, sets up the listener name and the list of events to subscribe to. */
+    public PatientOnCancelDeleter()
+    {
+        super("patient-oncancel-deleter", new PatientEditingCancelingEvent());
+    }
+
+    @Override
+    public void onEvent(final Event event, final Object source, final Object data)
+    {
+        Patient patient = ((PatientEvent) event).getPatient();
+        XWikiContext context = contextProvider.get();
+        XWiki wiki = context.getWiki();
+        try {
+            XWikiDocument doc = wiki.getDocument(patient.getDocument(), context);
+            if (isNew(doc, context)) {
+                wiki.deleteDocument(doc, true, context);
+
+                EntityReference defaultReference = new EntityReference(wiki.getDefaultPage(context),
+                    EntityType.DOCUMENT, new EntityReference(wiki.getDefaultSpace(context), EntityType.SPACE));
+                defaultReference = currentResolver.resolve(defaultReference, EntityType.DOCUMENT);
+                String url = wiki.getURL(new DocumentReference(defaultReference), "view", context);
+                context.getResponse().sendRedirect(url);
+            }
+        } catch (XWikiException ex) {
+            logger.error("Could not delete patient. {}", ex.getMessage());
+        } catch (IOException ex) {
+            logger.error("Could not redirect user appropriately. {}", ex.getMessage());
+        }
+    }
+
+    private boolean isNew(XWikiDocument patient, XWikiContext context)
+    {
+        // to be safe and not delete in case of error.
+        boolean hasVersion = false;
+        int versionCount = 0;
+        try {
+            List<String> versions = patient.getRevisions(new RevisionCriteria(), context);
+            for (String version : versions) {
+                XWikiRCSNodeInfo info = patient.getRevisionInfo(version, context);
+                if (!StringUtils.equalsIgnoreCase(info.getComment().trim(), AUTOSAVE_COMMENT) && !info.isMinorEdit()) {
```

Comment:
As discussed with @antonkatz the act of creating a patient adds a version, and clicking save adds another, so supposedly if ANY changes have been saved at least once there will be at least two non-minor versions. TODO: check what kind of edit is permission changes?